dialogue-core/src/main/java/com/palantir/dialogue/core/StickyEndpointChannels2.java

```diff
@@ -160,11 +148,33 @@ public void onFailure(Throwable _throwable) {
 
                 ListenableFuture<Response> callInFlightSnapshot = callInFlight;
                 if (callInFlightSnapshot == null) {
-                    ListenableFuture<Response> result = executeWithStickyToken(request, endpointChannel);
+                    ListenableFuture<Response> executeWithStickyTokenResult =
+                            executeWithStickyToken(request, endpointChannel);
                     // callInFlight must be updated prior to adding the callback, otherwise a quick completion
                     // may unset 'callInFlight' before it has been set in the first place!
+                    SettableFuture<Response> result = SettableFuture.create();
                     callInFlight = result;
-                    DialogueFutures.addDirectCallback(result, initialRequestCallback);
+                    DialogueFutures.addDirectCallback(executeWithStickyTokenResult, new FutureCallback<>() {
```

Comment:
Might be worth mentioning something about the ordering requirement here

Reply:
Done